File: src/main/java/com/walletwise/demo/controller/ExpenseController.java
package com.walletwise.demo.controller;

import com.walletwise.demo.dto.ApiResponse;
import com.walletwise.demo.dto.ExpenseRequest;
import com.walletwise.demo.dto.ExpenseResponse;
import com.walletwise.demo.entity.Expense;
import com.walletwise.demo.entity.User;
import com.walletwise.demo.repository.ExpenseRepository;
import com.walletwise.demo.service.UserService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/expenses")
public class ExpenseController {

    private static final Logger log = LoggerFactory.getLogger(ExpenseController.class);

    @Autowired
    private ExpenseRepository expenseRepository;

    @Autowired
    private UserService userService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<ExpenseResponse>>> getExpenses(
            @RequestHeader("Authorization") String token) {
        try {
            User user = userService.getUserByToken(token);
            List<Expense> expenses = expenseRepository.findByUserOrderByDateDesc(user);
            List<ExpenseResponse> responses = expenses.stream()
                    .map(e -> new ExpenseResponse(e.getId(), e.getTitle(), e.getAmount(), e.getCategory(), e.getDate(), e.getDescription(), e.getReceiptUrl()))
                    .collect(Collectors.toList());
            return ResponseEntity.ok(new ApiResponse<>(true, "Expenses fetched successfully", responses));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new ApiResponse<>(false, e.getMessage(), null));
        }
    }

    @PostMapping
    public ResponseEntity<ApiResponse<ExpenseResponse>> createExpense(
            @RequestHeader("Authorization") String token,
            @RequestBody ExpenseRequest request) {
        log.info("Received request to create expense with title: {}", request.getTitle());
        try {
            User user = userService.getUserByToken(token);
            Expense expense = new Expense();
            expense.setTitle(request.getTitle());
            expense.setAmount(request.getAmount());
            expense.setCategory(request.getCategory());
            expense.setDate(request.getDate());
            expense.setDescription(request.getDescription());
            expense.setReceiptUrl(request.getReceiptUrl());
            expense.setUser(user);

            expense = expenseRepository.save(expense);
            log.info("Expense record saved successfully to database. ID: {}", expense.getId());

            ExpenseResponse response = new ExpenseResponse(
                    expense.getId(),
                    expense.getTitle(),
                    expense.getAmount(),
                    expense.getCategory(),
                    expense.getDate(),
                    expense.getDescription(),
                    expense.getReceiptUrl()
            );

            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(new ApiResponse<>(true, "Expense created successfully", response));
        } catch (Exception e) {
            log.error("Failed to save expense record. Error: {}", e.getMessage(), e);
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, e.getMessage(), null));
        }
    }

    @RequestMapping(value = "/{id}", method = {RequestMethod.PATCH, RequestMethod.PUT})
    public ResponseEntity<ApiResponse<ExpenseResponse>> updateExpense(
            @RequestHeader("Authorization") String token,
            @PathVariable UUID id,
            @RequestBody ExpenseRequest request) {
        try {
            User user = userService.getUserByToken(token);
            Expense expense = expenseRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Expense entry not found"));

            if (!expense.getUser().getId().equals(user.getId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(new ApiResponse<>(false, "You do not have permission to modify this entry", null));
            }

            if (request.getTitle() != null) expense.setTitle(request.getTitle());
            if (request.getAmount() != null) expense.setAmount(request.getAmount());
            if (request.getCategory() != null) expense.setCategory(request.getCategory());
            if (request.getDate() != null) expense.setDate(request.getDate());
            if (request.getDescription() != null) expense.setDescription(request.getDescription());
            if (request.getReceiptUrl() != null) expense.setReceiptUrl(request.getReceiptUrl());

            expense = expenseRepository.save(expense);

            ExpenseResponse response = new ExpenseResponse(
                    expense.getId(),
                    expense.getTitle(),
                    expense.getAmount(),
                    expense.getCategory(),
                    expense.getDate(),
                    expense.getDescription(),
                    expense.getReceiptUrl()
            );

            return ResponseEntity.ok(new ApiResponse<>(true, "Expense updated successfully", response));
        } catch (Exception e) {
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, e.getMessage(), null));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> deleteExpense(
            @RequestHeader("Authorization") String token,
            @PathVariable UUID id) {
        try {
            User user = userService.getUserByToken(token);
            Expense expense = expenseRepository.findById(id)
                    .orElseThrow(() -> new RuntimeException("Expense entry not found"));

            if (!expense.getUser().getId().equals(user.getId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(new ApiResponse<>(false, "You do not have permission to delete this entry", null));
            }

            expenseRepository.delete(expense);
            return ResponseEntity.ok(new ApiResponse<>(true, "Expense deleted successfully", null));
        } catch (Exception e) {
            return ResponseEntity.badRequest()
                    .body(new ApiResponse<>(false, e.getMessage(), null));
        }
    }
}

File: walletwise-backend/src/controller/expense.controller.js
const expenseService = require('../service/expense.service');
const { sendResponse } = require('../utils/response.util');
const catchAsync = require('../utils/catchAsync');

exports.createExpense = catchAsync(async (req, res) => {
  let data = req.body;
  if (req.file) {
    data.receipt_image = req.file.path;
  }
  const expense = await expenseService.createExpense(req.user.id, data);
  sendResponse(res, 201, true, 'Expense added successfully', expense);
});

exports.updateExpense = catchAsync(async (req, res) => {
  let data = req.body;
  if (req.file) {
    data.receipt_image = req.file.path;
  }
  try {
    const expense = await expenseService.updateExpense(req.params.id, req.user.id, data);
    sendResponse(res, 200, true, 'Expense updated successfully', expense);
  } catch (err) {
    res.status(404);
    throw err;
  }
});

exports.deleteExpense = catchAsync(async (req, res) => {
  try {
    await expenseService.deleteExpense(req.params.id, req.user.id);
    sendResponse(res, 200, true, 'Expense deleted successfully');
  } catch (err) {
    res.status(404);
    throw err;
  }
});

exports.getExpenses = catchAsync(async (req, res) => {
  const expenses = await expenseService.getExpenses(req.user.id, req.query);
  sendResponse(res, 200, true, 'Expenses retrieved successfully', expenses);
});

exports.getExpenseById = catchAsync(async (req, res) => {
  try {
    const expense = await expenseService.getExpenseById(req.params.id, req.user.id);
    sendResponse(res, 200, true, 'Expense retrieved successfully', expense);
  } catch (err) {
    res.status(404);
    throw err;
  }
});

exports.getMonthlyExpenses = catchAsync(async (req, res) => {
  const { month, year } = req.query;
  if (!month || !year) {
    res.status(400);
    throw new Error('Month and year are required');
  }
  const expenses = await expenseService.getMonthlyExpenses(req.user.id, month, year);
  sendResponse(res, 200, true, 'Monthly expenses retrieved successfully', expenses);
});

exports.getExpensesByCategory = catchAsync(async (req, res) => {
  const expenses = await expenseService.getExpenses(req.user.id, { category: req.params.category });
  sendResponse(res, 200, true, 'Category expenses retrieved successfully', expenses);
});
